libs/runtime/ebpf_ring_buffer.c

```diff
@@ -224,56 +149,90 @@ _Must_inspect_result_ ebpf_result_t
 ebpf_ring_buffer_reserve(
     _Inout_ ebpf_ring_buffer_t* ring, _Outptr_result_bytebuffer_(length) uint8_t** data, size_t length)
 {
-    ebpf_result_t result;
-    ebpf_lock_state_t state = ebpf_lock_lock(&ring->lock);
-    ebpf_ring_buffer_record_t* record = _ring_buffer_acquire_record(ring, length);
-    if (record == NULL) {
-        result = EBPF_INVALID_ARGUMENT;
-        goto Done;
-    }
 
-    record->header.locked = 1;
-    MemoryBarrier();
+    for (;;) {
+        int64_t producer_offset = ReadAcquire64(&ring->producer_offset);
+        // Record points to the next record to allocate.
+        ebpf_ring_buffer_record_t* record =
+            (ebpf_ring_buffer_record_t*)(ring->shared_buffer + (producer_offset & ring->mask));
+
+        int64_t remaining_space = ring->length - (producer_offset - ReadAcquire64(&ring->consumer_offset));
+        long effective_length = (long)length + 4;
+        if (remaining_space < effective_length) {
+            return EBPF_NO_MEMORY;
+        }
+
+        // Check if locked.
+        if (record->size != 0) {
+            // If locked, pause then try again.
+            // Use _mm_pause() on x86 and __yield() on ARM.
+#if defined(_M_X64) || defined(_M_IX86)
+            _mm_pause();
+#else
+            __yield();
+#endif
+            continue;
+        }
+
+        if (InterlockedBitTestAndSetAcquire(&record->size, 31) != 0) {
+            continue;
+        }
+
+        // Check if the producer offset changed after we read it.
+        if (ReadAcquire64(&ring->producer_offset) != producer_offset) {
+            // Clear the lock bit
+            InterlockedAnd(&record->size, ~(EBPF_RING_BUFFER_RECORD_FLAG_LOCKED));
+            continue;
+        }
+
+        remaining_space = ring->length - (producer_offset - ReadAcquire64(&ring->consumer_offset));
+        if (remaining_space < effective_length) {
+            // Clear the lock bit
+            InterlockedAnd(&record->size, ~(EBPF_RING_BUFFER_RECORD_FLAG_LOCKED));
+            return EBPF_NO_MEMORY;
+        }
+
+        // Grab the pointer to the record.
+        *data = record->data;
 
-    *data = record->data;
-    result = EBPF_SUCCESS;
-Done:
-    ebpf_lock_unlock(&ring->lock, state);
-    return result;
+        WriteRelease(&record->size, effective_length | EBPF_RING_BUFFER_RECORD_FLAG_LOCKED);
+        int64_t new_producer_offset = producer_offset + effective_length;
+        WriteRelease64(&ring->producer_offset, new_producer_offset);
+        break;
+    }
+
+    return EBPF_SUCCESS;
 }
 
 _Must_inspect_result_ ebpf_result_t
 ebpf_ring_buffer_submit(_Frees_ptr_opt_ uint8_t* data)
 {
-    if (!data) {
+    ebpf_ring_buffer_record_t* record = CONTAINING_RECORD(data, ebpf_ring_buffer_record_t, data);
+    long size = ReadAcquire(&record->size);
```

Comment:
it's not clear why this has acquire semantics

Reply:
Need to investigate.